src/ResourceManager/Resources/Commands.Resources/ActiveDirectory/Models/ActiveDirectoryClient.cs

```diff
@@ -0,0 +1,818 @@
+                }
+                adObjects = GraphClient.Objects.GetObjectsByObjectIds(new GetObjectsParameters { ObjectIds = objectIds.GetRange(i, objectIdBatchCount), IncludeDirectoryObjectReferences = true });
+                result.AddRange(adObjects.Select(o => o.ToPSADObject()));
+            }
+            return result;
+        }
+
+        public PSADGroup GetGroupByDisplayName(string displayName)
+        {
+            var group = FilterGroups(new ADObjectFilterOptions() { SearchString = displayName });
+            if (group.Count() > 1)
+            {
+                throw new InvalidOperationException(string.Format(ProjectResources.MultipleGroupsWithDisplayNameFound, displayName));
+            }
+
+            if (group.Count() == 0)
+            {
+                throw new InvalidOperationException(string.Format(ProjectResources.GroupWithDisplayNameDoesntExist, displayName));
+            }
+
+            return group.FirstOrDefault();
+        }
+
+        public IEnumerable<PSADGroup> FilterGroups(ADObjectFilterOptions options, ulong first = ulong.MaxValue, ulong skip = 0)
+        {
+            if (!string.IsNullOrEmpty(options.Id))
+            {
+                try
+                {
+                    // use GetObjectsByObjectId to handle Redirects in the CSP scenario
+                    PSADGroup group = this.GetObjectsByObjectId(new List<string> { options.Id }).FirstOrDefault() as PSADGroup;
+                    if (group != null)
+                    {
+                        return new List<PSADGroup> { group };
+                    }
+                }
+                catch {  /* The group does not exist, ignore the exception */ }
+            }
+            else
+            {
+                Rest.Azure.OData.ODataQuery<ADGroup> odataQuery = null;
+                if (options.Mail != null)
+                {
+                    odataQuery = new Rest.Azure.OData.ODataQuery<ADGroup>(g => g.Mail == options.Mail);
+                }
+                else
+                {
+                    if (!string.IsNullOrEmpty(options.SearchString) && options.SearchString.EndsWith("*"))
+                    {
+                        options.SearchString = options.SearchString.TrimEnd('*');
+                        odataQuery = new Rest.Azure.OData.ODataQuery<ADGroup>(g => g.DisplayName.StartsWith(options.SearchString));
+                    }
+                    else
+                    {
+                        odataQuery = new Rest.Azure.OData.ODataQuery<ADGroup>(g => g.DisplayName == options.SearchString);
+                    }
+                }
+
+                return new GenericPageEnumerable<ADGroup>(
+                    delegate ()
+                    {
+                        return GraphClient.Groups.List(odataQuery);
+                    }, GraphClient.Groups.ListNext, first, skip).Select(g => g.ToPSADGroup());
+            }
+
+            return new List<PSADGroup>();
```

Comment:
So, this is a  bit difficult to follow.  I would suggest moving this return value into the upper if, and making this one

```c#
if(!string.IsNullOrEmpty(options.Id)) {
}
else if (options.Mail != null) {
}
else {
}
``` 

statement